File: camera_initial_pose.py
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D

def camera_initial_pose():
    # Define the radius of the sphere
    r = 5
    offset = 0

    # Create mesh grid for spherical coordinates (theta, phi)
    theta = np.linspace(0, 2 * np.pi, 50)  # Angle around the z-axis (azimuth)
    phi = np.linspace(0, np.pi / 2, 25)    # Angle from the z-axis (elevation, only upper half)

    Theta, Phi = np.meshgrid(theta, phi)

    # Parametric equations for the sphere in Cartesian coordinates (World frame coordinates)
    X_position_of_lens = r * np.cos(Phi) + offset
    Z_position_of_lens = r * np.sin(Phi) * np.sin(Theta)
    Y_position_of_lens = r * np.sin(Phi) * np.cos(Theta)

    # Adjust plot settings for better visualization
    fig = plt.figure()
    ax = fig.add_subplot(111, projection='3d')

    # Set the muted red color for the surface
    red_color = [0.8, 0.5, 0.5]  # Muted red RGB values
    ax.plot_surface(X_position_of_lens, Y_position_of_lens, Z_position_of_lens, color=red_color, edgecolor='none', alpha=0.5)
    
    # Add labels and title
    ax.set_xlabel('X')
    ax.set_ylabel('Y')
    ax.set_zlabel('Z')
    ax.set_title('Camera Positions')
    ax.set_box_aspect([1, 1, 0.6])  # Aspect ratio

    # # Camera body (cuboid)
    # z_body = np.array([-0.5, 0.5, 0.5, -0.5, -0.5, 0.5, 0.5, -0.5])  # X coordinates of the body
    # y_body = np.array([-0.3, -0.3, 0.3, 0.3, -0.3, -0.3, 0.3, 0.3])  # Y coordinates of the body
    # x_body = np.array([-0.2 + 1.5 + offset, -0.2 + 1.5 + offset, -0.2 + 1.5 + offset, -0.2 + 1.5 + offset,
    #                    0.2 + 1.5 + offset, 0.2 + 1.5 + offset, 0.2 + 1.5 + offset, 0.2 + 1.5 + offset])  # Z coordinates of the body

    # ax.plot_trisurf(x_body, y_body, z_body, color=[0.3, 0.3, 0.3])  # Plot the camera body (gray color)

    # Camera lens (cylinder)
    # z_lens, y_lens, x_lens = [arr.flatten() for arr in np.meshgrid(np.linspace(-0.15, 0.15, 30), np.linspace(-0.3, 0.3, 30))]  # Define a cylinder for the lens
    # x_lens *= 0.3  # Adjust the height of the cylinder
    # ax.plot_surface(x_lens + 1, y_lens, z_lens + offset, color=[0.1, 0.1, 0.1], edgecolor='none')  # Dark color for lens

    # Adjust the plot
    ax.view_init(30, 60)
    plt.axis('equal')
    #plt.show()
    return X_position_of_lens, Y_position_of_lens, Z_position_of_lens

File: camera_initial_pose.m

close all;
% Parameters
r = 5;             % Radius of the hemisphere
offset = -1;       % Offset for the dome center along the x-axis
n_points = 100;    % Number of points for the trajectory
z_level = r / 2;   % Set a constant z-level for the horizontal trajectory within the dome

% Create mesh grid for the hemispherical dome (theta, phi)
theta = linspace(0, 2*pi, 50);    % Azimuth angle
phi = linspace(0, pi/2, 25);      % Elevation angle (upper half only)
[Theta, Phi] = meshgrid(theta, phi);

% Parametric equations for the hemisphere in Cartesian coordinates
X_dome = r * cos(Phi) + offset;
Y_dome = r * sin(Phi) .* cos(Theta);
Z_dome = r * sin(Phi) .* sin(Theta);

% Create a horizontal cosine trajectory along the y-axis at z = z_level
x_traj =  ones(1, n_points);                 % Keep x constant
y_traj = (linspace(-r/2, r/8, n_points));                  % Vary y across the dome
z_traj = z_level+sin( 2*pi * y_traj / r);   % Sine wave along y-axis

% Plot the hemispherical dome
figure;
hold on;
surf(X_dome, Y_dome, Z_dome, 'FaceColor', [0.8, 0.5, 0.5], 'EdgeColor', 'none', 'FaceAlpha', 0.5);

% Plot the horizontal cosine trajectory inside the dome along y-axis
plot3(x_traj, y_traj, z_traj, 'b', 'LineWidth', 2, 'DisplayName', 'Horizontal Sine Trajectory');

% Additional plot settings
xlabel('X');
ylabel('Y');
zlabel('Z');
title('Horizontal Cosine Trajectory Inside Hemispherical Dome');
axis equal;
xlim([-r+offset, r+offset]);
ylim([-r, r]);
zlim([0, r]);

% Show legend and grid
legend('show');
grid on;
hold off;
